crates/bevy_xr_utils/src/tracking_utils.rs

```diff
@@ -0,0 +1,314 @@
+            PreUpdate,
+            sync_actions
+                .before(OxrActionSetSyncSet)
+                .run_if(session_running),
+        );
+        //attach sets
+        app.add_systems(XrSessionCreated, attach_set);
+        //create actions
+        app.add_systems(Startup, create_actions.run_if(session_available));
+        //left grip
+        // .add_systems(
+        //     PreUpdate,
+        //     update_space_transforms
+        //         .in_set(OxrSpaceSyncSet)
+        //         .run_if(session_running),
+        // )
+        app.add_systems(PreUpdate, update_left_grip.after(update_space_transforms));
+        app.add_systems(PreUpdate, update_right_grip.after(update_space_transforms));
+    }
+}
+
+//stage
+fn update_stage(
+    mut root_query: Query<&mut Transform, (With<XrTrackingRoot>, Without<XRTrackedStage>)>,
+    mut stage_query: Query<&mut Transform, (With<XRTrackedStage>, Without<XrTrackingRoot>)>,
+) {
+    let tracking_root_transform = root_query.get_single_mut();
+    match tracking_root_transform {
+        Ok(root) => {
+            for (mut transform) in &mut stage_query {
+                *transform = root.clone();
+            }
+        }
+        Err(_) => (),
+    }
+}
+
+//view
+#[derive(Component)]
+struct HeadXRSpace(XrReferenceSpace);
+
+fn update_head_transforms(
+    session: Res<OxrSession>,
+    default_ref_space: Res<XrPrimaryReferenceSpace>,
+    pipelined: Option<Res<Pipelined>>,
+    frame_state: Res<OxrFrameState>,
+    mut query: Query<(&mut Transform, &HeadXRSpace, Option<&XrReferenceSpace>)>,
+) {
+    for (mut transform, space, ref_space) in &mut query {
+        let ref_space = ref_space.unwrap_or(&default_ref_space);
+        let time = if pipelined.is_some() {
+            openxr::Time::from_nanos(
+                frame_state.predicted_display_time.as_nanos()
+                    + frame_state.predicted_display_period.as_nanos(),
+            )
+        } else {
+            frame_state.predicted_display_time
+        };
+        let space_location = session.locate_space(&space.0, ref_space, time);
+
+        if let Ok(space_location) = space_location {
+            let flags = OxrSpaceLocationFlags(space_location.location_flags);
+            if flags.pos_valid() {
+                transform.translation = space_location.pose.position.to_vec3();
+            }
+            if flags.rot_valid() {
+                transform.rotation = space_location.pose.orientation.to_quat();
+            }
+        }
+    }
+}
+
+fn update_view(
+    mut head_query: Query<&mut Transform, (With<HeadXRSpace>, Without<XRTrackedView>)>,
+    mut view_query: Query<&mut Transform, (With<XRTrackedView>, Without<HeadXRSpace>)>,
+) {
+    let head_transform = head_query.get_single_mut();
+    match head_transform {
+        Ok(root) => {
+            for (mut transform) in &mut view_query {
+                *transform = root.clone();
+            }
+        }
+        Err(_) => (),
+    }
+}
```

Comment:
i am a little confused by this function, why not add the `XrTrackedView` component the entity in head_query directly?

Reply:
this one is unnecessary, I added the layer of redirection because I plan on eventually adding recording and playback.

Reply:
couldn't you just not insert the space when playing back a recording?